plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/TrinoS3ClientFactory.java

```diff
@@ -149,7 +162,49 @@ private static AWSCredentialsProvider getAwsCredentialsProvider(Configuration co
             return getCustomAWSCredentialsProvider(conf, providerClass);
         }
 
-        return DefaultAWSCredentialsProviderChain.getInstance();
+        AWSCredentialsProvider provider = getAwsCredentials(conf)
+                .map(value -> (AWSCredentialsProvider) new AWSStaticCredentialsProvider(value))
+                .orElseGet(DefaultAWSCredentialsProviderChain::getInstance);
+
+        String iamRole = conf.get(S3_IAM_ROLE);
+        if (iamRole != null) {
+            String stsEndpointOverride = conf.get(S3_STS_ENDPOINT);
+            String stsRegionOverride = conf.get(S3_STS_REGION);
+            String s3RoleSessionName = conf.get(S3_ROLE_SESSION_NAME);
+            String externalId = conf.get(S3_EXTERNAL_ID);
+
+            AWSSecurityTokenServiceClientBuilder stsClientBuilder = AWSSecurityTokenServiceClientBuilder.standard()
+                    .withCredentials(provider);
+
+            String region;
+            if (!isNullOrEmpty(stsRegionOverride)) {
+                region = stsRegionOverride;
+            }
+            else {
+                DefaultAwsRegionProviderChain regionProviderChain = new DefaultAwsRegionProviderChain();
+                try {
+                    region = regionProviderChain.getRegion();
+                }
+                catch (SdkClientException ex) {
+                    log.warn("Falling back to default AWS region " + US_EAST_1);
```

Comment:
nit: slight inconsistency between two code snippets

https://github.com/trinodb/trino/blob/64b0b7d9bf5b0314e0de64a8cee946bfeb283d99/plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java#L1143

Is there any chance to DRY / group together the common code used to create the provider?

Reply:
I will refactor to reuse `TrinoS3FileSystem.java` and will get rid of this `TrinoS3ClientFactory.java`.